### ISSUE TITLE
Integrate Asset Server - [merged]

### ISSUE THREAD

**Issue by benloh** [theRAPTLab/gsgo#601]
_Merges dev-bl/save-assets -> dev-next_

Branch: `dev-bl/save-assets`

Four significant changes: 

1. This moves assets out of `gem-srv/src/app/static/` into a local asset server at `gsgo/gs_assets` and sets the groundwork for a internet server.

2. New projects can now be created from project templates.

3. All project data is now saved both in the database and read/written as project files.  Project files can be saved in an arbitrary folder hierarchy under `/gsgo/gsgo-assets/`, provided you set the project root path.

4. The project data file format has changed slightly.  

* It requires valid JSON (e.g. keys need to be quoted)...
* ...with the exception of scripts, which allow template literals
* Metadata has been moved around
* Round parameters have been moved to metadata
* Various key names have been changed (including "label", "bpid", "scriptText")

Your old project file will not work as-is.  It will need to be converted to the new format.

---

# Project Files Overview

Project data is now read and saved as separate project files.  When a project is loaded, it is read from a `*.gemprj` file and loaded into the database.  Any changes to the project (e.g. editing project metadata, rounds, blueprints, and instances) are all automatically written to both the database and project file.

(There is a one second delay in writing to queue keystrokes and reduce the frequency of writing).

#### Project Root

Project files are all saved in an arbitrary 'root' folder under `gsgo/gs_assets`.  Only one 'root' folder is opened by the GEMSTEP app.  By default, the root is `local`.

Inside the root folder should be two folders:
* `projects`
* `sprites`

e.g. by default most servers should have the following folders:
```
gsgo/gs_assets/local/projects
gsgo/gs_assets/local/sprites
```

The 'root' folder can be named anything and placed anywhere under `gsgo/gs_assets`.  You can take advantage of this to arrange folders however is most convenient for you.  Use this to manage sets of projects for different locations/dates/sites/groups/studies/etc.  

For example, if you have a single server machine that will be used in 'room13' and 'room21', but when you're in one room, you only want to see projects from that room, you can arrange projects like this:

```
gsgo/gs_assets/room13
gsgo/gs_assets/room21
```

On the other hand, if you wanted to nest folders, you could also do something like this:
```
gsgo/gs_assets/orgs/iu/fall2021/morning
gsgo/gs_assets/orgs/iu/fall2021/afternoon
gsgo/gs_assets/orgs/iu/fall2021/serverB
gsgo/gs_assets/orgs/iu/winter2021/all
gsgo/gs_assets/orgs/vu/sept2021/day1
gsgo/gs_assets/orgs/vu/sept2021/day1
```
Each of these folders would be a project root and contain any number of project files and sprites.

`gsgo-settings.js` defines the project root that will be used. (In the future there will be a local override version of this file so you don't have to worry about inadvertently changing other core settings.)

To set the project root, set the `GS_ASSETS_PROJECT_ROOT` variable, e.g.

* To use `gsgo/gs_assets/room13`:

  ```const GS_ASSETS_PROJECT_ROOT = 'room13';```

* To use `gsgo/gs_assets/orgs/iu/fall2021/morning`:

  ```const GS_ASSETS_PROJECT_ROOT = 'orgs/iu/fall2021/morning';```

NOTE: No trailing `/` is needed.

For convenience you can always just leave the `local` folder setting and just stuff all of your projects there.

#### `projects` folder and project names

Inside the 'root' folder, you always need to define a `projects` folder and a `sprites` folder.  The `projects` folder contains all the individual project files.  The `sprites` folder contains sprites shared across all the projects.

Project filenames need to be unique.  So if you have multiple groups working on similar files, you would name them according to the group.  For example:

```
gsgo/gs_assets/room13/projects/aquatic_grp1.gemprj
gsgo/gs_assets/room13/projects/aquatic_grp2.gemprj
gsgo/gs_assets/room13/projects/aquatic_blue.gemprj
gsgo/gs_assets/room13/projects/aquatic_green.gemprj
```

Project filenames are now used as the URL parameter for the project.  For example, `gsgo/gs_assets/room13/projects/aquatic_grp1.gemprj` can be opened via:

```http://localhost/app/project?project=aquatic_grp1```

#### `sprites` folder

There is no longer a need to add and maintain a `assets.json` file.  Asset lists are now generated automatically.

The `sprites` folder should contain all the spritesheet definitions (json) and graphics (png) files shared by all projects.

```
gsgo/gs_assets/room13/sprites/algae.json
gsgo/gs_assets/room13/sprites/algae.png
gsgo/gs_assets/room13/sprites/bee.json
gsgo/gs_assets/room13/sprites/bee.png
```

#### Project Templates

New projects can be created by creating a new project from a project template file.  Use this to define a base project for students to work on.  They would then create a new project based on the template and modify that new project.

Any project file can be converted into a template file by adding a `_template_` prefix to the name.

So a typical curriculum development workflow might be something like this:
1. You work on `aquatic.gemprj` until you feel like it's at a good starting point for students.
2. Rename `aquatic.gemprj` to `_template_aquatic.gemprj`
3. When a student logs onto the system and wants to create a new project, they will see the template aquatic file.
4. When they select the template, they will be prompted to enter a new filename.

#### Project Files Read/Write Import/Export

Projects are ALWAYS read from the `*.gemprj` files.  So you can easily just add more `*.gemprj` files to the project root, or replace any existing ones.  Even if a project file was previously opened and stored in the database, when you open it again, it will be read fresh from the `*.gemprj` file and saved to the database.

#### `*.gemprj` format

The `*.gemprj` format is basically a modified JSON format.  

* You can edit the `*.gemprj` with a text editor if needed.  
* You should be able to set your text editor to treat it as a JSON or javascript file to get syntax highlighting.
* To make the files more editable we convert script lines into template literals with carriage returns.  This technically makes the project file invalid JSON.
* All other text should be double-quoted, including key names.

#### Duplicating / Save As

Changing the id of a project in the Project Settings editor UI will cause the app to save the project with a new name matching the new id.  The old project file will remain.  This is effectively a duplicate/save as feature.

---

# To Test

Due to all the system level changes, you will have to do a bit of work to prep for this version.

You MUST do this to prep.  NOTE these instructions are very different from prior merges, so follow them carefully.  Do NOT skip a step!

#### Set up local assets

Do this to verify that the local assets are working.

1. `git fetch&&git checkout dev-bl/save-assets`
2. Remove existing `db.loki` file at `gsgo/gs_packages/gem-srv/runtime/db.loki` so a new updated database will be created.
3. Create a local asset server:

    1. Create a `gs_assets` folder at the root level of your `gsgo` repo -- you should have a folder at `gsgo/gs_assets`.
    2. If you don't have it already, clone a local `art-assets` repo: https://gitlab.com/stepsys/gem-step/art-assets somewhere else on your hard disk.
    3. In the `art-assets` repo, switch to the `asset-server` branch: `git fetch&&git checkout asset-server`
    4. Copy the `art-assets` folder to `gsgo/gs_assets`.  So you should end up with `gsgo/gs_assets/art-assets` folder.  Inside that should be a `sprites` folder with about 80 items (`.json` and `.png` files).
    5. Rename `art-assets` to `local` (the default project root is `local`).  You should end up with `gsgo/gs_assets/local` and `gsgo/gs_assets/local/projects` and `gsgo/gs_assets/local/sprites` folders

4. Run `npm ci`
5. Run `npm run bootstrap`
6. Run `npm start`
7. Go to `http://localhost/app/main?project=decomposition` -- The project should load with all art assets.
8. On a separate computer, load a Viewer or CharController and make sure you can see the art assets.

#### Try Loading a Different Project

You should be able to open other projects.

1. Go to Login: `http://localhost/app/login`
2. Select another project.
3. Select Main -- Project should load with all assets.

#### Verify Saving Works

Modifying project settings, round settings, blueprints, and instances should all trigger saves to the project.

Try modifying something and make sure the project file reflects the changes.

#### Create a new Project file from Project Template

Students will typically create a new project from an existing template.

1. Go to Login: `http://localhost/app/login`
2. Click the "New Aquatic Ecosystems" button to create a new project.
3. The system will prompt you for a name, suggesting a new name with a random number suffix.  During a study you probably would want students to attach their group name or something like that.
4. Click OK.
5. Click "Main" to open the Main window.
6. Note that the URL has `project=aquatic_<your suffix>`.  You can always use this to open the file.
7. Aquatic should load with all art assets
8. Modify something.
9. Check the project file in `gsgo/gs_assets/local/projects` and make sure the changes were saved.

---

# Background

## Long Term Goal vs Interim Plan

The eventual goal for asset management is to provide a permanent asset server (e.g. hosted on Digital Ocean) that any project can use to download assets.  This merge request is one step in that direction, but is not there yet.

#### How a Permanent Host might work

* A Digital Ocean instance would run an asset server script, e.g. `npm run asset`
* Assets that are meant to be shared are stored in a `gs_assets_hosted` folder at the root level of the server.
* Any project can then reference those assets via a url, e.g. `http://123.456.789:8080/assets/moths/predator`
* The project would download those assets and cache them locally.

There are still a few issues to work out before we implement a permanent host, though (e.g. how to upload and maintain assets on the server, how assets might need to be divided/organized by group or location or project).  So this is probably still a few weeks away.

#### How Interim Local Host works

This merge request focuses on integrating the new asset manager system with local hosting.

* The local asset host is defined in `/gsgo-settings.js` (or a local override that will be developed later).
* Currently it is set to use `localhost` as the local host URL.
* Assets are put in an `gs_assets` folder.
* Using the `art-assets` repo, we
  * Place `art-assets` in `gsgo/gs_assets`
  * Rename `art-assets/local` to `art-assets/sprites` -- `sprites` is where the asset manager looks for sprites by default.
* There is no longer a need to create and maintain an `assets.json` file.  The manifest file is now automatically created.
* Plain `*.png` no longer work currently.  You need to define a corresponding `*.json` file to define bounds.

#### Daily Use

Once this is merged into dev-next, here's how you would work with assets:

** Adding/Updating Assets **
If you need to add or update assets, you would need to make sure you:
* Update assets in the `gsgo/gs_assets/art-assets/sprites` folder
* Commit any changes to the `art-assets` repo's `asset-server` branch (for now at least...once we fully transition to the assets server we can merge that into the main branch).

---

# Changes

* Assets are no longer in `gsgo/gs_packages/gem-srv/src/app/static/`, but moved to the top level in `gsgo/gs_assets`.
* The `assets.json` file is no longer necessary.  Manifests are now auto-generated.
* `png`-only assets are no longer supported.  You need to create a `.json` file.  (This will be addressed with #329 )

---
